distsys/resources/filesystem.go

```diff
@@ -0,0 +1,85 @@
+package resources
+
+import (
+	"fmt"
+	"github.com/UBC-NSS/pgo/distsys"
+	"io/ioutil"
+	"path"
+)
+
+// FilesystemArchetypeResourceMaker produces a distsys.ArchetypeResourceMaker for a filesystem-backed
+// map-like resource. Each element of the map will refer to a file, with keys and values being required
+// to be string-typed, and keys being required to refer to valid paths (or create-able paths, if a
+// key is written to before it is read).
+func FilesystemArchetypeResourceMaker(workingDirectory string) distsys.ArchetypeResourceMaker {
+	return IncrementalArchetypeMapResourceMaker(func(index distsys.TLAValue) distsys.ArchetypeResourceMaker {
+		return distsys.ArchetypeResourceMakerFn(func() distsys.ArchetypeResource {
+			return &fileArchetypeResource{
+				workingDirectory: workingDirectory,
+				subPath:          index.AsString(),
+			}
+		})
+	})
+}
+
+type fileArchetypeResource struct {
+	distsys.ArchetypeResourceLeafMixin
+
+	workingDirectory string
+	subPath          string
+
+	writePending *string
+	cachedRead   *string
+}
+
+var _ distsys.ArchetypeResource = &fileArchetypeResource{}
+
+func (res *fileArchetypeResource) Abort() chan struct{} {
+	res.writePending = nil
+	res.cachedRead = nil
+	return nil
+}
+
+func (res *fileArchetypeResource) PreCommit() chan error {
+	return nil
+}
+
+func (res *fileArchetypeResource) Commit() chan struct{} {
+	if res.writePending != nil {
+		doneCh := make(chan struct{})
+		go func() {
+			err := ioutil.WriteFile(path.Join(res.workingDirectory, res.subPath), []byte(*res.writePending), 0777)
```

Comment:
If we want to make sure that the file is really persisted, we should call `fsync` system call but I think `ioutil.WriteFile` only calls `write` system call. AFAIK, `write` syscall has a buffer that persisted periodically using `fsync`.

Reply:
Well, if you want full atomicity, I think the only way is to write a temp file them `os.Move` it over the original (the move being the only trustworthy atomic filesystem op).

This code isn't even tested (only the read part is), so maybe it makes more sense to leave a TODO and come back to do it properly? (I can add the TODO)

Reply:
Adding TODO is fine I think. However, I'm not sure about `os.Move` since it seems a bit weird.

Reply:
Maybe multiple things are needed; that's perhaps why this should be left as an additional change. Atomic file ops hard.

For reference, I was trying to remember this package when talking about `os.Move` (which I remembered incorrectly, it should have been `os.Rename`): https://github.com/natefinch/atomic

Reply:
Didn't about that. Thanks for the link.